omod/src/main/java/org/openmrs/module/reportingrest/web/controller/ReportingRestController.java

```diff
@@ -28,11 +57,139 @@ public class ReportingRestController extends MainResourceController {
 
     public static final String REPORTING_REST_NAMESPACE = "/reportingrest";
 
+    private static final Log LOGGER = LogFactory.getLog(ReportingRestController.class);
+
     /**
      * @see org.openmrs.module.webservices.rest.web.v1_0.controller.BaseRestController#getNamespace()
      */
     @Override
     public String getNamespace() {
         return RestConstants.VERSION_1 + REPORTING_REST_NAMESPACE;
     }
+
+    @RequestMapping(value = "/runReport", method = RequestMethod.POST)
+    @ResponseStatus(HttpStatus.OK)
+    public void runReport(@RequestBody RunReportRequest runReportRequest) {
+        ReportDefinition reportDefinition = Context.getService(ReportDefinitionService.class)
+            .getDefinitionByUuid(runReportRequest.getReportDefinitionUuid());
+
+        ReportService reportService = getReportService();
+
+        Map<String, Object> parameterValues = new HashMap<String, Object>();
+        for (Parameter parameter : reportDefinition.getParameters()) {
+            Object convertedObj =
+                convertParamValueToObject(runReportRequest.getReportParameters().get(parameter.getName()), parameter.getType());
+            parameterValues.put(parameter.getName(), convertedObj);
+        }
+
+        List<RenderingMode> renderingModes = reportService.getRenderingModes(reportDefinition);
+        RenderingMode renderingMode = null;
+        for (RenderingMode mode : renderingModes) {
+            if (StringUtils.equals(mode.getArgument(), runReportRequest.getRenderModeUuid())) {
+                renderingMode = mode;
+                break;
+            }
+        }
+
+        final ReportRequest reportRequest;
+        if(StringUtils.isNotBlank(runReportRequest.getExistingRequestUuid())) {
+            reportRequest = reportService.getReportRequestByUuid(runReportRequest.getExistingRequestUuid());
+        } else {
+            reportRequest = new ReportRequest();
+        }
+        reportRequest.setReportDefinition(new Mapped<ReportDefinition>(reportDefinition, parameterValues));
+        reportRequest.setRenderingMode(renderingMode);
+        reportRequest.setPriority(ReportRequest.Priority.NORMAL);
+        reportRequest.setSchedule(runReportRequest.getSchedule());
+
+        reportService.queueReport(reportRequest);
+        reportService.processNextQueuedReports();
+    }
+
+    @RequestMapping(value = "/cancelReport", method = RequestMethod.DELETE)
+    @ResponseStatus(HttpStatus.OK)
+    public void cancelReport(@RequestParam String reportRequestUuid) {
+        ReportService reportService = getReportService();
+        ReportRequest reportRequest = reportService.getReportRequestByUuid(reportRequestUuid);
+        if (reportRequest != null) {
+            reportService.purgeReportRequest(reportRequest);
+        }
+    }
+
+    @RequestMapping(value = "/preserveReport", method = RequestMethod.POST)
```

Comment:
How different is `preserveReport` from `saveReport`?

Reply:
@dkayiwa I'll change it to `saveReport`, preserve came from UI. 

@mseaton Generally, this is a quick migration of legacy UI's `org.openmrs.module.reportingui.fragment.controller.ReportStatusFragmentController`, this directly relates to actions on UI.

The existing `ReportRequestResource` is a `CrudResource` - should I just add an endpoint method there?
Or do you mean, that if there is an update call, which changes request's status from COMPLETED to SAVED, then it should run the `ReportService.savereport()`?